### PR TITLE
Use BinaryProvider for building

### PR DESCRIPTION
**Pull request by SimonDanisch**
- [x] Windows
- [x] OSX (untested)
- [x] Linux X11
- [x] Linux Wayland ( I think cmake should pick that up)

https://github.com/glfw/glfw/issues/1297

### CODE REVIEW

**Comment by SimonDanisch** [comment]
Should we try a hybrid solution, with BinaryProvider for OSX & Windows, and keep using the old infrastucture for Linux? Should already be an improvement, since OSX with homebrew was likely the one failing the most. For windows we probably won't get any improvement... just an improvement in consistency I guess ;)

File: deps/build.jl [hunk 1]
```diff
@@ -1,59 +1,36 @@
...
-	provides(Homebrew.HB, "glfw", glfw, os=:Darwin)
-end
-
-# download a pre-compiled binary (built by GLFW)
-if iswindows()
-	archive = "glfw-$version.bin.WIN$(Sys.WORD_SIZE)"
-	libpath = joinpath(archive, "lib-mingw-w64")
-	uri = URI("https://github.com/glfw/glfw/releases/download/$version/$archive.zip")
-	provides(Binaries, uri, glfw, unpacked_dir=archive, installed_libpath=libpath, os=:Windows)
+using BinaryProvider # requires BinaryProvider 0.3.0 or later
```
Comment (jayschwa) [inline deps/build.jl]: If this requires a minimum version of `BinaryProvider`, then please encode that in `REQUIRE` file.

File: deps/build.jl [hunk 2]
```diff
@@ -1,59 +1,36 @@
...
+
+# Parse some basic command-line arguments
+const verbose = "--verbose" in ARGS
+const prefix = Prefix(get([a for a in ARGS if a != "--verbose"], 1, joinpath(@__DIR__, "usr")))
+products = [
+    LibraryProduct(prefix, "", :lib),
+]
+
+# Download binaries from hosted location
+bin_prefix = "https://github.com/SimonDanisch/GLFWBuilder/releases/download/3.2.1"
```
Comment (jayschwa) [inline deps/build.jl]: This project should probably be moved under the JuliaGL org once it's stable.

File: deps/build.jl [hunk 3]
```diff
@@ -1,59 +1,36 @@
...
+
+# Listing of files generated by BinaryBuilder:
+download_info = Dict(
+    Windows(:i686) => ("$bin_prefix/glfw.v1.0.0.i686-w64-mingw32.tar.gz", "08bdec95b829c46f5aee3d98d4ba4a29eb79c0553f5ec68749d926598fa1f4b8"),
+    MacOS(:x86_64) => ("$bin_prefix/glfw.v1.0.0.x86_64-apple-darwin14.tar.gz", "7b4c712c0e194c7196750b14d3178ba8ece8eeeed0fbac64f3b6e90579f0efda"),
```
Comment (jayschwa) [inline deps/build.jl]: Why is the version in these files 1.0.0?

Comment (jayschwa) [inline deps/build.jl]: What does the `14` in the `darwin14` denote? Do you know how many versions of OS X this build will work for?

File: src/GLFW.jl [hunk 4]
```diff
@@ -3,7 +3,11 @@ __precompile__()
 module GLFW
 
 include("compat.jl")
-include("../deps/deps.jl")
+if isfile(joinpath(@__DIR__, "..", "deps", "deps.jl"))
+	include("../deps/deps.jl")
+else
+	error("GLFW wasn't build. Please run Pkg.build(\"GLFW\")")
```
Comment (jayschwa) [inline src/GLFW.jl]: Typo: "built"

**Comment by Gnimuc** [comment]
Shall we revisit the hybrid solution?

File: deps/build.jl (hunk 3 above)
Comment (SimonDanisch) [reply deps/build.jl]: no idea^^

**Comment by SimonDanisch** [comment]
I finally added the hybrid solution..Hope this builds fine on all platforms! Please test this :)

File: REQUIRE [hunk 5]
```diff
@@ -1,4 +1,3 @@
 julia 0.7
-BinDeps 0.8.10
-@!windows CMakeWrapper
-@osx Homebrew 0.7.0
+BinaryProvider 0.3.0
+CMake
```
Comment (Gnimuc) [inline REQUIRE]: better to drop `REQUIRE` and add `Project.toml` instead.

Comment (SimonDanisch) [reply REQUIRE]: should be in another PR

**Comment by SimonDanisch** [comment]
Anyone with OSX ready to try this out?

**Comment by asinghvi17** [comment]
On MacOS v10.14.4:

```julia
(v1.1) pkg> add GLFW#sd/binarybuilder
  Updating git-repo `https://github.com/JuliaGL/GLFW.jl.git`
 Resolving package versions...
 Installed Tables ────────────── v0.1.19
 Installed LowRankApprox ─────── v0.2.3
 Installed PDMats ────────────── v0.9.7
 Installed ImageAxes ─────────── v0.6.0
 Installed BlockBandedMatrices ─ v0.4.1
 Installed InfiniteArrays ────── v0.1.0
  Updating `~/.julia/environments/v1.1/Project.toml`
  [f7f18e0c] ↑ GLFW v2.3.0 ⇒ v2.3.0+ #sd/binarybuilder (https://github.com/JuliaGL/GLFW.jl.git)
  Updating `~/.julia/environments/v1.1/Manifest.toml`
  [ffab5731] ↑ BlockBandedMatrices v0.4.0 ⇒ v0.4.1
  [f7f18e0c] ↑ GLFW v2.3.0 ⇒ v2.3.0+ #sd/binarybuilder (https://github.com/JuliaGL/GLFW.jl.git)
  [2803e5a7] ↑ ImageAxes v0.5.0 ⇒ v0.6.0
  [4858937d] ↑ InfiniteArrays v0.0.3 ⇒ v0.1.0
  [898213cb] ↑ LowRankApprox v0.2.2 ⇒ v0.2.3
  [90014a1f] ↑ PDMats v0.9.6 ⇒ v0.9.7
  [bd369af6] ↑ Tables v0.1.18 ⇒ v0.1.19
  Building GLFW → `~/.julia/packages/GLFW/NRJVQ/deps/build.log`

(v1.1) pkg> build GLFW
  Building CMake → `~/.julia/packages/CMake/ZGawK/deps/build.log`
  Building GLFW ─→ `~/.julia/packages/GLFW/NRJVQ/deps/build.log`
```

**Comment by SimonDanisch** [comment]
@jayschwa, any other comments, or you do you think we can merge this?

File: deps/build.jl [hunk 6]
```diff
@@ -1,46 +1,84 @@
...
+        rm(tarball_dir, force=true, recursive=true)
+        rm(source_path, force=true, recursive=true)
+        mkpath(dirname(tarball_dir))
+        unpack(tarball_path, dirname(tarball_dir); verbose=true)
+        tarpath = joinpath(dirname(tarball_dir), "glfw-$(version)")
+        mv(tarpath, source_path, force = true)
+        build_dir = joinpath(source_path, "build")
+        mkdir(build_dir)
+        cd(build_dir) do
+            opts = ["-DBUILD_SHARED_LIBS=on", "-DGLFW_BUILD_DOCS=Off", "-DGLFW_BUILD_EXAMPLES=off"]
```
Comment (jayschwa) [inline deps/build.jl]: Tests may need to be disabled too.

File: deps/build.jl [hunk 7]
```diff
@@ -1,46 +1,84 @@
...
+    end
+else
 
-# get library through Homebrew package manager
-@static if Sys.isapple()
-	using Homebrew
-	provides(Homebrew.HB, "glfw", glfw, os=:Darwin)
-end
+    # Download binaries from hosted location
+    bin_prefix = "https://github.com/SimonDanisch/GLFWBuilder/releases/download/5.0.0"
```
Comment (jayschwa) [inline deps/build.jl]: The `GLFWBuilder` repo should be moved to the `JuliaGL` organization?

File: REQUIRE (hunk 5 above)
Comment (jayschwa) [reply REQUIRE]: Let's start CMake with its latest released version (1.1.1, it looks like).

**Pull request merged**